CMakeLists.txt

```
@@ -0,0 +1,225 @@
+    target_link_libraries(mold PRIVATE
+        ZLIB::ZLIB
+        ${M_LIBRARIES})
+endif()
+
+target_link_libraries(mold PRIVATE
+    TBB::tbb
+    $<$<BOOL:MOLD_USE_MIMALLOC>:mimalloc::mimalloc>)
+
+if(CMAKE_SYSTEM_NAME MATCHES Linux)
+    find_package(OpenSSL REQUIRED COMPONENTS Crypto)
+    target_link_libraries(mold PRIVATE OpenSSL::Crypto)
+endif()
+
+if(CMAKE_SYSTEM_PROCESSOR MATCHES riscv64
+    OR CMAKE_SYSTEM_PROCESSOR MATCHES armv6)
+    find_library(ATOMIC_LIBRARIES atomic REQUIRED)
+    target_link_libraries(mold PRIVATE ${ATOMIC_LIBRARIES})
+endif()
+
+file(STRINGS .git/HEAD GIT_HASH)
+
+if(GIT_HASH MATCHES "ref: ")
+    string(REPLACE "ref: " "" GIT_HASH "${GIT_HASH}")
+    file(STRINGS .git/${GIT_HASH} GIT_HASH)
+endif()
+
+target_compile_definitions(mold PRIVATE
+    GIT_HASH="${GIT_HASH}"
+    MOLD_VERSION="${CMAKE_PROJECT_VERSION}")
+
+target_sources(mold PRIVATE
+    compress.cc
+    demangle.cc
+    filepath.cc
+    glob.cc
+    hyperloglog.cc
+    main.cc
+    multi-glob.cc
+    perf.cc
+    strerror.cc
+    tar.cc
+    uuid.cc
+    elf/arch-arm32.cc
+    elf/arch-arm64.cc
+    elf/arch-i386.cc
+    elf/arch-riscv64.cc
+    elf/arch-x86-64.cc
+    elf/cmdline.cc
+    elf/dwarf.cc
+    elf/gc-sections.cc
+    elf/icf.cc
+    elf/input-files.cc
+    elf/input-sections.cc
+    elf/linker-script.cc
+    elf/lto.cc
+    elf/main.cc
+    elf/mapfile.cc
+    elf/output-chunks.cc
+    elf/passes.cc
+    elf/relocatable.cc
+    elf/subprocess.cc
+    macho/arch-arm64.cc
+    macho/arch-x86-64.cc
+    macho/cmdline.cc
+    macho/dead-strip.cc
+    macho/input-files.cc
+    macho/input-sections.cc
+    macho/lto.cc
+    macho/main.cc
+    macho/mapfile.cc
+    macho/output-chunks.cc
+    macho/tapi.cc
+    macho/yaml.cc
+    third-party/rust-demangle/rust-demangle.c)
+
+# Create the ld and ld64 symlinks required for testing
+add_custom_command(
+    TARGET mold POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E create_symlink mold ld
+    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    COMMENT "Creating ld symlink"
+    VERBATIM)
+add_custom_command(
+    TARGET mold POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E create_symlink mold ld64
+    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    COMMENT "Creating ld64 symlink"
+    VERBATIM)
+
+include(CTest)
+
+set(MOLD_TESTS_PATH ${CMAKE_CURRENT_LIST_DIR}/test/)
+
+if(CMAKE_SYSTEM_NAME MATCHES Darwin)
+    string(APPEND MOLD_TESTS_PATH "macho")
+else()
+    string(APPEND MOLD_TESTS_PATH "elf")
+endif()
+
+string(APPEND MOLD_TESTS_PATH "/*")
+FILE(GLOB MOLD_TESTS ${MOLD_TESTS_PATH})
+
+foreach(MOLD_TEST IN LISTS MOLD_TESTS)
+    add_test(NAME ${MOLD_TEST} COMMAND ${MOLD_TEST})
+endforeach()
+
+# FIXME(cwasser): This is just a convenience ctest target,
+# which ensures that mold is build. This should ideally be fixed in CMake.
+include(ProcessorCount)
+ProcessorCount(MOLD_NPROC)
+add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} -j${MOLD_NPROC}
+    DEPENDS mold USES_TERMINAL)
```

"fixed in CMake"? CMake needs no fixing in this regard, you just need to let go of Linux-isms. You can use presets to automate testing to the point of needing to only run `ctest --preset=dev`.

The `DEPENDS` argument is also not how things work https://cmake.org/cmake/help/latest/command/add_custom_target.html#:~:text=at%20build%20time.-,DEPENDS,-Reference%20files%20and

---

I wouldn't call this Linux-isms. When I run `cargo test` on one of my Rust projects, it ensures that the binaries, required for testing, are actually build first and not fail with an obscure error about missing files...

---

Ninja and the Makefiles generators already have the `test` target, while other generators have `RUN_TESTS`. This target is created when `enable_testing()` or `include(CTest)` is used to enable testing. This target depends on "all" and makes sure everything is built.

Is this what you mean?

---

Hmm okay, the target doesn't actually depend on the "all" target. The `check` target could still use improvement:

```suggestion
include(ProcessorCount)
ProcessorCount(MOLD_NPROC)
add_custom_target(
    check
    COMMAND "${CMAKE_COMMAND}" --build "${mold_BINARY_DIR}" --config "\$<CONFIG>" -j "${MOLD_NPROC}"
    COMMAND "${CMAKE_CTEST_COMMAND}" -C "\$<CONFIG>" -j "${MOLD_NPROC}"
    VERBATIM
    USES_TERMINAL
)
```

---

Thanks!

---

I've integrated this target and it works great, thanks! :smile: